Unix-domain sockets (AF_LOCAL)

mikbras opened the pull request:
This PR implements Unix-domain sockets in which all data is exchanged within the kernel (inside the enclave).

### Limitations:

- No support for ``sendmsg()`` with ``MSG_PEEK`` (see https://github.com/deislabs/mystikos/issues/1083).
- No support for control data (see https://github.com/deislabs/mystikos/issues/1084).
- No support for ``dup()`` (https://github.com/deislabs/mystikos/issues/1085).

### Approach

The Unix-domain sockets implementation exchanges data within the kernel but creates a host socket pair for the purposes of managing file-descriptor events (read-write enablement). The implementation is similar to Mystikos pipes but accommodates full-duplex traffic.

### Syslog

This PR introduces a rudimentary **syslog** feature, whereby the kernel can print **syslog** messages to standard error in red, yellow, and green (depending on the severity).

This feature is used to print warnings when an app attempts to use unsupported Unix-domain socket features, including:
- Creating sockets with an unsupported socket domain (e.g., ``AF_ALG``)
- Creating sockets with unsupported socket types (e.g., ``AF_SOCK_RAW``)
- Sending control data with ``sendmsg()``
- Peeking at Unix-domain socket data (``recvmsg()`` with ``MSG_PEEK``)
- Calling ``setsockopt()`` with unsupported ``optname`` arguments.

There may be other unsupported scenarios, which might be uncovered by more testing.

jxyang commented on include/myst/sockdev.h:
```diff
@@ -163,4 +163,14 @@ struct myst_sockdev
 
 myst_sockdev_t* myst_sockdev_get(void);
 
+myst_sockdev_t* myst_sockdev_get_uds(void);
```
Suggest renaming to `myst_udsdev_get` or `myst_sockdev_uds_get`.

mikbras replied:
Good idea. I went with ``myst_udsdev_get()``.

mikbras commented on kernel/udsdev.c:
```diff
@@ -0,0 +1,1797 @@
+    myst_list_t list;
+} acceptor_t;
+
+struct myst_sock
```
``struct myst_sock`` is defined in ``<myst/sockdev.h>`` and both ``sockdev.c`` and ``usddev.c`` define their own local definitions of the structure (same for file systems with ``struct myst_file``).

mikbras commented on utils/hex.c:
```diff
@@ -22,6 +22,28 @@ void myst_hexdump(const char* label, const void* data, size_t size)
+        if (c >= ' ' && c <= '~')
+            printf("%c", c);
+        else
+            printf("<%x>", c);
```
Good idea. Fixed.

mikbras commented on tests/sockets/sockets.c:
```diff
@@ -44,7 +89,7 @@ static void* _srv_thread_func(void* arg)
     (void)arg;
 
-    assert((lsock = socket(args->domain, SOCK_STREAM, 0)) >= 0);
+    assert((lsock = socket(args->domain, SOCK_STREAM | SOCK_CLOEXEC, 0)) >= 0);
```
Oops, that was experimental. I reverted this change.

mikbras commented on tests/cpython-tests/config.json:
```diff
@@ -9,7 +9,7 @@
 
     // The heap size of the user application. Increase this setting if your app experienced OOM.
-    "MemorySize": "8g",
+    "MemorySize": "1g",
```
Ooops, I didn't mean to leave that change in. I reverted it.

mikbras commented on kernel/sockdev.c:
```diff
@@ -805,3 +811,118 @@ extern myst_sockdev_t* myst_sockdev_get(void)
+            if (!(type & SOCK_STREAM) && !(type & SOCK_DGRAM))
+            {
+                const char* str = myst_socket_type_str(type);
+                MYST_WLOG("unsupported socket type: %d: %s", type, str);
```
Fixed. These will be printed in **red** now. :)

mikbras commented on kernel/udsdev.c:
```diff
@@ -0,0 +1,1797 @@
+        myst_mutex_destroy(&sock->mutex);
+        myst_buf_release(&sock->buf);
+        memset(sock, 0, sizeof(myst_sock_t));
+        free(sock);
```
Linux leaves the Unix-domain socket file after after closing the listener socket. I wrote this little test program to verify.

```
#include <assert.h>
#include <netinet/in.h>
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

int main()
{
    int sockfd;
    assert((sockfd = socket(AF_LOCAL, SOCK_STREAM, 0)) >= 0);

    struct sockaddr_un addr;
    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    strcpy(addr.sun_path, "/tmp/my_uds");

    assert(bind(sockfd, (struct sockaddr*)&addr, sizeof(addr)) == 0);

    close(sockfd);

    /* "/tmp/my_uds" survives */

    return 0;
}
```

mikbras commented:
Thank you Xuejun for looking at this. I addressed all your review comments pretty much as requested. One thing: Linux leaves the Unix-domain socket after closing the listener socket (I wrote a test for it to be sure).
